pricegraph/src/orderbook.rs

```diff
@@ -255,18 +384,124 @@ mod tests {
         let order = orderbook.orders.best_order_for_pair(pair).unwrap();
         assert_eq!(orderbook.num_orders(), 3);
         assert_eq!(order.user, user_id(3));
-        assert!(num::eq(
-            order.weight(),
-            orderbook.get_projected_pair_weight(pair),
-        ));
+        assert_approx_eq!(order.weight(), orderbook.get_projected_pair_weight(pair));
 
         orderbook.update_projection_graph();
         let order = orderbook.orders.best_order_for_pair(pair).unwrap();
         assert_eq!(orderbook.num_orders(), 1);
         assert_eq!(order.user, user_id(1));
-        assert!(num::eq(
-            order.weight(),
-            orderbook.get_projected_pair_weight(pair),
-        ));
+        assert_approx_eq!(order.weight(), orderbook.get_projected_pair_weight(pair));
+    }
+
+    #[test]
+    fn fills_path_and_updates_projection() {
+        //             /---1.0---v
+        // 0 --1.0--> 1          2 --1.0--> 3 --1.0--> 4
+        //             \---2.0---^                    /
+        //                        \--------1.0-------/
+        //
+        // NOTE: Higher prices are worse.
```

Comment:
I might be wrong, but I would have thought it's the other way around. The numerator is the buyAmount and the denominator the sellAmount (https://github.com/gnosis/dex-contracts/blob/ca45b5d0fd2cbfc46cc14631e16d700dce2b2c57/contracts/BatchExchange.sol#L570). The more you sell the smaller the fraction (price) becomes and at the the same time the worse off you are (you are selling more while receiving the same).

Reply:
So higher prices are worse for pushing an amount through the graph, but are better for the owners of the order. That means that, when filling a path through the graph with the lowest cost, we want the lowest price (so highest sell amount) since they benefit us (the one pushing a flow through the graph) more by virtue of using existing orders that are worse off for the people who placed them.

Reply:
So the point of view matters, the higher prices are worse for the person trying to place an order that takes advantage of the existing orders.

Reply:
What I meant by:
> So the point of view matters, the higher prices are worse for the person trying to place an order that takes advantage of the existing orders.

Is that if we are trying to answer the question "what is the best value for `x` for my order: buy 1A for at most `x`B", then we do so by examining all orders that are "buy at most `b`B for at least `a`A" so that we can "push" our B token to A.

Given this, as you said, the best order for us is a small as possible `x` value (so the price goes up), and in order to achieve that, we need minimize the price of the reciprocal orders (hence the smaller price is better for getting a lower cost pushing some token through the graph) which is to say, minimize `b/a`, so higher `a`, the sell amount in this case, is better for our point of view.

Reply:
I think I see it now. Maybe you can emphasize in the note that this is from the point of view of someone trying to push flow against the existing orders.